### DISCUSSION TITLE
How to output the optimized reflactance

### DISCUSSION THREAD

**Discussion by EMwavelength** [mitsuba-renderer/mitsuba3#196] [Q&A] [answered]
hi,
   First of all i am a complete beginner in this area so please guide me, i am running mitsuba 3 inverse rendering example and successfully completed the cornell box example now my problem is how to output the recovered reflectance textures so they can be re-used, i am sure this is the purpose of inverse rendering to recover the reflectance of real world objects for later use cases. So is there a  way in mitsuba 3 to output the  optimized texture such as recovered red reflectance texture i.e diffuse albedo.

**Comment by merlinND** [comment, marked as answer]
Hello @EMwavelength,

I've converted your Issue to a Discussion because it doesn't concern a bug in the system.

In your optimization script, you've probably noticed a `params` object (acting like a dictionary) that contains all of the parameters being optimized.

Let's say the name of the parameter containing a diffuse albedo texture is called `'wall.bsdf.data'`. Then you can get an image out with something like:

```python
img = mi.Bitmap(params['wall.bsdf.data'])
img.write('optimized_texture.exr')
```

(didn't try to run this particular code, let me know if this doesn't work).

In the case of the Cornell Box in the inverse rendering tutorial, it doesn't use a texture but a single RGB value, so I think you could just print it after the optimization is done:

```python
for k, v in params.items():
    print(k, v)
```

**Comment by EMwavelength** [comment]
Hi,
   Sorry for confusion i meant textures not just rgb values, and thank you successfully recovered the reflectance of scene.First i imported the scene from 3ds max then given the target and initial image i recovered the results.I downgraded the the recovered texture from 2k to
512 resolution for uploading. The recovered texture was of same size as of source texture.

Initial:
![initial](https://user-images.githubusercontent.com/111644839/185790784-fc9b905e-53a6-421e-a223-393a9bd324bb.jpg)
Target
![target](https://user-images.githubusercontent.com/111644839/185790800-271ef8f8-cef7-4765-88d6-34bc0ff85fc0.jpg)
Optimized
![optimized](https://user-images.githubusercontent.com/111644839/185790807-986a4dab-91c6-45f8-bea4-a22b8c0a63d4.jpg)
Optimized texture
![optimized_texture](https://user-images.githubusercontent.com/111644839/185790814-22d7538c-3908-4912-9967-9c535839c55a.jpg)

**Reply by merlinND** [reply]
Great to hear this works!
Note the quality of the recovered texture can certainly be improved with more samples per pixel, lowering the learning rate toward the end of the optimization, etc.

**Reply by EMwavelength** [reply]
Thank you figured out here the much more better output:

Improved texture:
![improved_texture](https://user-images.githubusercontent.com/111644839/185792540-433d4534-89fb-49d8-8c43-d65dfdad3f46.jpg)
